File: services/GeneralService.java
package com.example.JohnCouponPart2.services;
import com.example.JohnCouponPart2.Repositories.CouponRepository;
import com.example.JohnCouponPart2.Repositories.CustomerRepository;
import com.example.JohnCouponPart2.beans.Category;
import com.example.JohnCouponPart2.beans.Coupon;
import com.example.JohnCouponPart2.beans.Customer;
import com.example.JohnCouponPart2.exceptions.CouponException;
import com.example.JohnCouponPart2.exceptions.CustomerException;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class GeneralService {

    private CustomerRepository customerRepository;
    private CouponRepository couponRepository;

    public GeneralService(CustomerRepository customerRepository, CouponRepository couponRepository) {
        this.customerRepository = customerRepository;
        this.couponRepository = couponRepository;
    }

    public Customer customerRegister (Customer customer) throws CustomerException {
        // Check if a customer with the same email already exists
        if (!customerRepository.existsByEmail(customer.getEmail())) {
            // Save the new customer
            customerRepository.save(customer);
            return customer;
        } else {
            throw new CustomerException("Email is already in use.");
        }
    }

    public List<Coupon> getAllCoupons(){
        return couponRepository.findAll();
    }

    public List<Coupon> getAllCouponsByCategory(Category category){
        return couponRepository.findCouponsByCategory(category);
    }

    public List<Coupon> getAllCouponsByPrice(double price){
        return couponRepository.findCouponsByPriceLessThanEqual(price);
    }

    public List<Coupon> getAllCouponsByCategoryAndPrice(Category category, double price){
        return couponRepository.findCouponsByPriceLessThanEqualAndCategory(price, category);
    }

    public Coupon getOneCoupon(int id) throws CouponException {
        return couponRepository.findById(id).orElseThrow(() -> new CouponException("Coupon doesn't exist!"));

    }
}

File: src/Services/GeneralService.ts
import axios from "axios";
import Coupon from "../Models/Coupon";
import Category from "../Models/Category";
import Customer from "../Models/Customer";

class GeneralService {
    public async getAllCoupons() {
        const response = await axios.get<Coupon[]>("http://localhost:8080/getcoupons");
        return response.data;
    }

    public async getAllCouponsByCategory(category: Category) {
        const response = await axios.get<Coupon[]>(`http://localhost:8080/getcouponsbycategory?category=${category}`);
        return response.data;
    }

    public async getAllCouponsByPrice(price: number) {
        const response = await axios.get<Coupon[]>(`http://localhost:8080/getcouponsbyprice?price=${price}`);
        return response.data;
    }

    public async getAllCouponsByCategoryAndPrice(category: Category, price: number) {
        const response = await axios.get<Coupon[]>(`http://localhost:8080/getcouponsbycategoryandprice?category=${category}&price=${price}`);
        return response.data;
    }

    public async getOneCoupon(id: number) {
        const response = await axios.get<Coupon>(`http://localhost:8080/coupondetails/${id}`);
        return response.data;
    }

    public async customerRegister(customer:Customer){
        const response = await axios.post<Customer>("http://localhost:8080/register", customer)
        return response.data
    }
}

const generalService = new GeneralService();
export default generalService;
